Mathlib/GroupTheory/SpecificGroups/KleinFour.lean

```diff
@@ -39,74 +40,195 @@ produces the third one.
+class IsAddKleinFour (G : Type*) [AddGroup G] : Prop where
+  card_four : Nat.card G = 4
+  exponent_two : AddMonoid.exponent G = 2
+
+/-- A Klein four-group is a group of cardinality four and exponent two. -/
+@[to_additive existing IsAddKleinFour]
+class IsKleinFour (G : Type*) [Group G] : Prop where
+  card_four : Nat.card G = 4
+  exponent_two : Monoid.exponent G = 2
+
+attribute [simp] IsKleinFour.card_four IsKleinFour.exponent_two
+  IsAddKleinFour.card_four IsAddKleinFour.exponent_two
+
+instance : IsAddKleinFour (ZMod 2 × ZMod 2) where
+  card_four := by simp
+  exponent_two := by simp [AddMonoid.exponent_prod]
+
+instance : IsKleinFour (DihedralGroup 2) where
+  card_four := by simp only [Nat.card_eq_fintype_card]; rfl
+  exponent_two := by simp [DihedralGroup.exponent]
+
+instance {G : Type*} [Group G] [IsKleinFour G] :
+    IsAddKleinFour (Additive G) where
+  card_four := by rw [← IsKleinFour.card_four (G := G)]; congr!
+  exponent_two := by simp
+
+instance {G : Type*} [AddGroup G] [IsAddKleinFour G] :
+    IsKleinFour (Multiplicative G) where
+  card_four := by rw [← IsAddKleinFour.card_four (G := G)]; congr!
+  exponent_two := by simp
+
+namespace IsKleinFour
+
+@[to_additive]
+instance instFinite {G : Type*} [Group G] [IsKleinFour G] : Finite G :=
+  Nat.finite_of_card_ne_zero <| by norm_num [IsKleinFour.card_four]
+
+@[to_additive (attr := simp)]
+lemma card_four' {G : Type*} [Group G] [Fintype G] [IsKleinFour G] :
+    Fintype.card G = 4 :=
+  Nat.card_eq_fintype_card (α := G).symm ▸ IsKleinFour.card_four
+
+open Finset
+
+variable {G : Type*} [Group G]
+
+/-- A group of order 4 is not cyclic if and only if its exponent is two. -/
+@[to_additive]
+lemma not_isCyclic_iff (hG : Nat.card G = 4) :
+    ¬ IsCyclic G ↔ Monoid.exponent G = 2 := by
+  let _inst : Fintype G := @Fintype.ofFinite G <| Nat.finite_of_card_ne_zero <| by norm_num [hG]
+  have hG' : Fintype.card G = 4 := by simpa using hG
+  refine ⟨fun h_cyc ↦ ?_, fun h_exp h_cyc ↦ by simpa [hG', h_exp] using h_cyc.exponent_eq_card⟩
+  apply le_antisymm
+  · refine Nat.le_of_dvd zero_lt_two <| Monoid.exponent_dvd_of_forall_pow_eq_one G 2 fun g ↦ ?_
+    have hg' : orderOf g ∈ {1, 2, 4} :=
+      (Nat.mem_divisors (m := 4)).mpr ⟨hG' ▸ orderOf_dvd_card, by norm_num⟩
+    simp only [mem_singleton, Nat.succ.injEq, OfNat.one_ne_ofNat, mem_insert, or_self,
+      orderOf_eq_one_iff] at hg'
+    obtain (rfl|hg|hg) := hg'
+    · simp
+    · exact hg ▸ pow_orderOf_eq_one g
+    · apply False.elim <| h_cyc <| isCyclic_of_orderOf_eq_card g (hG' ▸ hg)
+  · have := (Fintype.one_lt_card_iff_nontrivial (α := G)).mp <| by norm_num [hG']
+    exact Monoid.one_lt_exponent
+
+variable [IsKleinFour G]
+
+@[to_additive]
+lemma not_isCyclic : ¬ IsCyclic G :=
+  fun h ↦ by let _inst := Fintype.ofFinite G; simpa using h.exponent_eq_card
+
+@[to_additive]
+lemma inv_eq_self (x : G) : x⁻¹ = x := inv_eq_self_of_exponent_two (by simp) x
```

Comment:
If `IsKlein` is a class this should be `simp`. Similarly below.

Reply:
I think that's a bad simp lemma for a prop mixin class. Any time lean sees an inverse it will trigger TC synthesis for `IsKleinFour`

Reply:
Fair enough. So let's add a comment explaining that this shouldn't be a global simp-lemma, but that it can be enabled locally when working with a Klein four group.